### PR TITLE
feat: lifecycle script execution and allowBuilds policy

### PR DESCRIPTION
**Pull request by Saturate**
## Summary

- Add lifecycle script runner to `pacquet-executor`: preinstall, install, postinstall in correct order, `binding.gyp` auto-detection, `INIT_CWD`/`PNPM_SCRIPT_SRC_DIR`/PATH environment setup
- Add `allowBuilds` build policy matching pnpm 11's default-deny behavior, with `dangerouslyAllowAllBuilds` escape hatch
- Wire build step into the frozen-lockfile install path
- Port `buildSequence` + `graphSequencer` so packages with `requires_build` run children-first
- Add `pnpm:lifecycle` and `pnpm:ignored-scripts` reporter channels — script start/stdio/exit events flow through the same NDJSON pipeline `@pnpm/cli.default-reporter` consumes from upstream pnpm
- Port 11 upstream lifecycle script tests as `known_failures`

Upstream refs:
- [`runPostinstallHooks`](https://github.com/pnpm/pnpm/blob/7e91e4b35f/exec/lifecycle/src/index.ts)
- [`runLifecycleHook`](https://github.com/pnpm/pnpm/blob/80037699fb/exec/lifecycle/src/runLifecycleHook.ts)
- [`createAllowBuildFunction`](https://github.com/pnpm/pnpm/blob/7e91e4b35f/building/policy/src/index.ts)
- [`buildSequence`](https://github.com/pnpm/pnpm/blob/80037699fb/building/during-install/src/buildSequence.ts)
- [`graphSequencer`](https://github.com/pnpm/pnpm/blob/80037699fb/deps/graph-sequencer/src/index.ts)
- [`lifecycleLogger`](https://github.com/pnpm/pnpm/blob/80037699fb/core/core-loggers/src/lifecycleLogger.ts)
- [`ignoredScriptsLogger`](https://github.com/pnpm/pnpm/blob/80037699fb/core/core-loggers/src/ignoredScriptsLogger.ts)

## What's implemented

- `run_postinstall_hooks` in `pacquet-executor` faithfully ports pnpm's hook runner. The hook is now generic over `R: Reporter` and emits `pnpm:lifecycle` events through the standard pacquet reporter pipeline: `Script` before each spawn, one `Stdio` event per stdout/stderr line read through `Stdio::piped()` byline buffering, and `Exit` after the script returns
- `AllowBuildPolicy` reads `pnpm.allowBuilds` and `pnpm.dangerouslyAllowAllBuilds` from `package.json` with correct tri-state semantics (allow/deny/ignored)
- Default-deny matches pnpm 11: packages must be explicitly listed in `allowBuilds` to run their scripts
- `BuildModules::run::<R>` runs lifecycle scripts for `requires_build: true` packages in the frozen-lockfile path and returns the sorted (peer-stripped) list of packages whose scripts were skipped because they weren't in `allowBuilds`. `InstallFrozenLockfile::run` emits one `pnpm:ignored-scripts` event with that list (always, even when empty), mirroring upstream
- `build_sequence` walks the dep graph from each importer, keeps only nodes whose subtree contains a build node, and feeds that subgraph into `graph_sequencer` to produce topologically ordered chunks. `BuildModules` iterates the chunks in order so children build before parents
- `AllowBuildPolicy` and `BuildModules.lockfile_dir` use the existing `requester` field (derived from `manifest.path().parent()`) instead of `std::env::current_dir()`
- New `LogEvent::Lifecycle` and `LogEvent::IgnoredScripts` variants with wire-shape tests
- Recording-fake reporter tests for `run_postinstall_hooks` (Script→Stdio→Exit ordering, non-zero exit propagation) and for `BuildModules::run` (sorted ignored-builds, explicit-deny exclusion). Existing `install_emits_pnpm_event_sequence` updated to expect the new `IgnoredScripts` event between `Stats` and `ImportingDone`
- Unit tests for the build policy, key parsing, `graph_sequencer`, and `build_sequence`

## What's not yet exercisable end-to-end

All 11 integration tests are `known_failures` because:
1. The non-frozen install path does not write a lockfile, so tests cannot exercise `--frozen-lockfile` yet
2. Bin linking (#330) is not implemented, so scripts that invoke dependency binaries fail

Once either lockfile writing or pre-generated lockfile fixtures land, the tests that use simple `node` scripts (no dependency bins) can be promoted. The rest need pnpm/pacquet#330.

## Known limitations / follow-ups

The remaining gaps vs upstream pnpm are tracked in pnpm/pacquet#397, including missing `npm_*` lifecycle env vars, ancestor-`node_modules/.bin` PATH walking, `linkBinsOfDependencies` before scripts, Windows shell selection, optional-dep failure tolerance, `patchedDependencies`, `isBuilt`/side-effects-cache skipping, build concurrency, and a few minor items.

This PR also still has:
- `ignoreScripts` config flag — not yet wired
- `pendingBuilds`/`ignoredBuilds` persistence in `.modules.yaml` — not yet wired
- Build chunks run sequentially — pnpm runs each chunk concurrently bounded by `childConcurrency` (perf-only)

## Test plan

- [x] `cargo test -p pacquet-package-manager build_modules::tests` (15 unit tests, including the new ignored-builds tests)
- [x] `cargo test -p pacquet-package-manager build_sequence::tests` (5 unit tests)
- [x] `cargo test -p pacquet-package-manager graph_sequencer::tests` (8 unit tests)
- [x] `cargo test -p pacquet-executor lifecycle::tests` (3 unit tests, recording-fake reporter)
- [x] `cargo test -p pacquet-reporter lifecycle_event ignored_scripts` (2 wire-shape tests)
- [x] `cargo test -p pacquet-cli --test lifecycle_scripts` (11 known_failures)
- [x] `just ready`
- [x] `taplo format --check`

Covers: pnpm/pnpm#11633 (Stage 1, "Support building dependencies")

---
Written by an agent (Claude Code, claude-opus-4-7).

## Summary by CodeRabbit

# Release Notes

* **New Features**
  * Added lifecycle script execution support for dependencies (preinstall, install, postinstall) during installation with proper build ordering
  * Added `pnpm.allowBuilds` configuration to selectively enable or disable lifecycle scripts per package

* **Tests**
  * Added comprehensive test coverage for lifecycle script execution scenarios and build dependency management

[Review Change Stack](https://app.coderabbit.ai/change-stack/pnpm/pacquet/pull/391)

### CODE REVIEW

File: crates/cli/tests/lifecycle_scripts.rs [hunk 1]
```diff
@@ -0,0 +1,531 @@
...
+        fs::remove_dir_all(&node_modules).expect("remove node_modules");
+
+        eprintln!("Running pacquet install --frozen-lockfile...");
+        CommandTempCwd::init()
+            .add_mocked_registry()
+            .pacquet
+            .with_current_dir(&workspace)
+            .with_args(["install", "--frozen-lockfile"])
+            .assert()
+            .success();
```
[the remark this answers is not in the document]
Comment (zkochan) [reply crates/cli/tests/lifecycle_scripts.rs]: Resolved differently from the suggested patch: the inner `mock_instance` is dropped, but `AutoMockInstance::load_or_init` returns an anchor to a shared singleton process, and the outer `mock_instance` from line 153 is held alive through `drop((root, mock_instance, frozen_root))` at the end of the test, so the registry stays running. The `frozen_root` binding additionally keeps the inner `TempDir` alive through the command. Fixed in d6fdf96.

---
Written by an agent (Claude Code, claude-opus-4-7).

**Comment by KSXGitHub** [comment]
`ignoredBuilds` should be covered by <https://github.com/pnpm/pacquet/pull/332>. I am actively reviewing this. I hope this PR will complete soon enough.

Bin linking should be covered by <https://github.com/pnpm/pacquet/pull/333>.

**Comment by KSXGitHub** [comment]
Also, the DI pattern as documented in <https://github.com/pnpm/pacquet/issues/339> and as implemented in my 3 PRs have helped plugging all coverage holes in <https://github.com/pnpm/pacquet/pull/332>.

**Comment by zkochan** [comment]
@KSXGitHub keep in mind that this is not yet production code. It doesn't need to be perfect as long as the benchmarks are good. It is OK to ship ugly code faster and file an issue for future refactor.

**Comment by KSXGitHub** [comment]
@zkochan 

> It is OK to ship ugly code faster and file an issue for future refactor.

Well. "Nothing is more permanent than a temporary solution" or so they say.

Also. Strict tests and strict code are important in this age of AI code generation. They make it less likely for AI to introduce bugs and subtle mistakes.

Fast iteration and quick prototyping is important, yes, but we're no longer writing code manually, do we? The cost of perfect code is on the AI whilst the toll of imperfect code is on the human reviewers.

That is what I think.

**Comment by Saturate** [comment]
Thanks for the context on pnpm/pacquet#339, @KSXGitHub. The DI pattern makes sense for unit-testing error paths in `AllowBuildPolicy::from_manifest` and `run_postinstall_hooks` (both currently call `std::fs::read_to_string` and `Command::new` directly).

Since pnpm/pacquet#339 hasn't landed and the final names are still being decided, this PR uses direct calls for now. Happy to refactor to the DI pattern once the conventions are settled and shipped to main.

**Comment by KSXGitHub** [comment]
@Saturate So the DI pattern actually makes sense here?

Then just apply it. For now, name it `Api` (trait), `RealApi` (struct), and stuffs.

> Happy to refactor to the DI pattern once the conventions are settled and shipped to main.

"Shipped to main" means that there shall be a refactor PR to touch many parts of the code, including this one.

Anyway, just do DI now. If the names are unfit, they shall be renamed later in a different PR.

**Comment by KSXGitHub** [comment]
The DI issue also has the "When dependency injection is not needed" section as well. The AI implementing this should take it into consideration.

I haven't read this code in details to know whether DI would actually be required.

**Comment by KSXGitHub** [comment]
The wording from <https://github.com/pnpm/pacquet/pull/391#issuecomment-4403919587> seems to suggest that some error paths are impossible to test normally. If so, DI makes sense.

**Comment by zkochan** [comment]
@KSXGitHub the postinstall tests in the pnpm/pnpm repository don't use any mocking. They test real packages with install scripts from the registry-mock and the public registry. So, I am not sure why dependency injection is needed here.

**Comment by KSXGitHub** [comment]
@zkochan We are porting with AI. AIs make mistakes.

DI is only used for unreachable edge cases.

Those edge cases are unreachable even in pnpm/pnpm.

**Comment by zkochan** [comment]
> The DI issue also has the "When dependency injection is not needed" section as well. The AI implementing this should take it into consideration.
>
> I haven't read this code in details to know whether DI would actually be required.

The instructions are to do exact copy of pnpm/pnpm including tests. If there are edge cases that are not covered in pnpm, there is no requirement to cover them in the rust rewrite.

**Comment by zkochan** [comment]
@coderabbitai full review

**Review by zkochan — APPROVED** [review]
